Program.cs

```diff
@@ -582,44 +595,71 @@ public OperationOutcome ValidateWithJava()
-      Task<OperationOutcome> validateWithJava = Task.Run(() => ValidateWithJava());
-      // .ContinueWith(System.Threading.Tasks.Task <OperationOutcome> t =>
-      // {
-      //   setOutcome(t.Result, ValidatorType.Java);
-      //   ValidatingJava = false;
-      // });
-      // TaskScheduler.FromCurrentSynchronizationContext()
-      Task<OperationOutcome> validateWithDotnet = Task.Run(() => ValidateWithDotnet());
+      Task<OperationOutcome> validateWithJava = Task.Run(() => ValidateWithJava(token), token);
+      Task<OperationOutcome> validateWithDotnet = Task.Run(() => ValidateWithDotnet(token), token);
 
       var allTasks = new List<Task> { validateWithJava, validateWithDotnet };
       while (allTasks.Any())
       {
-        var finished = await Task.WhenAny(allTasks);
-        if (finished == validateWithJava)
-        {
-          allTasks.Remove(validateWithJava);
-          var result = await validateWithJava;
-          SetOutcome(result, ValidatorType.Java);
-          ValidatingJava = false;
-        }
-        else if (finished == validateWithDotnet)
-        {
-          allTasks.Remove(validateWithDotnet);
-          var result = await validateWithDotnet;
-          SetOutcome(result, ValidatorType.Dotnet);
-          ValidatingDotnet = false;
-        }
-        else
-        {
-          allTasks.Remove(finished);
-        }
+        try {
+          var finished = await Task.WhenAny(allTasks);
+          if (finished == validateWithJava)
+          {
+            allTasks.Remove(validateWithJava);
+            var result = await validateWithJava;
+            SetOutcome(result, ValidatorType.Java);
+            ValidatingJava = false;
+          }
+          else if (finished == validateWithDotnet)
+          {
+            allTasks.Remove(validateWithDotnet);
+            var result = await validateWithDotnet;
+            SetOutcome(result, ValidatorType.Dotnet);
+            ValidatingDotnet = false;
+          }
+          else
+          {
+            allTasks.Remove(finished);
+          }
+        } catch (OperationCanceledException) {
+          // When we signalled to cancel the validation, the
+          // OperationCanceledException is thrown whenever we await the task.
+          // This prevents processing the results, effectively decoupling the
+          // task. We don't need to handle the exception itself.
+		    }
       }
     }
+
+    public void CancelValidation()
+    {
+      // Signal the CancellationToken in the tasks that we want to cancel.
+      if (validatorCancellationSource != null) {
+        validatorCancellationSource.Cancel();
+        validatorCancellationSource.Dispose();
+      }
+      validatorCancellationSource = null;
+      
+      // We can actively kill the Java validator as this is an external
+      // process. The .NET validator needs to run its course until completion,
+      // we'll just ignore the results.
+      foreach (Process process in validatorProcesses) {

```

Comment:
```suggestion
      foreach (var process in validatorProcesses) {
```

Redundant to qualify it here